File: 프로그래머스/lv2/42746. 가장 큰 수/가장 큰 수.js
function solution(numbers) { // numbers = [3, 30, 34, 5, 9]
    let strings = numbers.map((num) => String(num)); // strings = ['3', '30', '34', '5', '9']
    strings.sort((a, b) => (b + a) - (a + b));       // strings = ['9', '5', '34', '3', '30']
    // 만약 a = '3', b = '30'일 때
    // (b + a) - (a + b) = '303' - '330' = -27로, 
    //  330이 앞에 오고, 303이 뒤에 오는 것이 더 큰 수를 만들 수 있다.
    
    // 만약 a = '30', b = '34일' 때
    // (b + a) - (a + b) = '3430' - '3034' = 396으로, 
    // 34가 앞에 오고, 30이 뒤에 오는 것이 더 큰 수를 만들 수 있다.
    
    const answer = strings.join(''); // '9534330'
    
    return answer[0] === "0" ? "0" : answer; // 만약 answer가 '0____'으로 시작하면 0을 리턴하도록 조건
}

File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.py
# def solution(numbers):
#     sorted_numbers = bubble_sort(numbers)
#     stringified_numbers = map(str, sorted_numbers) # map(lambda x: str(x), sorted_numbers)
    
#     return ''.join(stringified_numbers)

# def bubble_sort(numbers):
#     n = len(numbers)
    
#     for i in range(n):
#         for j in range(n - i - 1): # 매번 루프가 끝날 때마다, 맨 오른쪽 값은 이미 정렬된 상태이므로 비교 범위를 i만큼 줄여야 함
#             # print(j, ' 와 ', j + 1, ' 을 비교')
#             if str(numbers[j]) + str(numbers[j + 1]) < str(numbers[j + 1]) + str(numbers[j]):
#                 # 스왑
#                 numbers[j], numbers[j + 1] = numbers[j + 1], numbers[j]

#     return numbers

# 버블 정렬은 시간 복잡도가 O(n^2)이므로 O(nlogn)인 파이썬 내장 메서드 sorted() 사용하는 게 나음
def solution(numbers):
    sorted_numbers = sorted(numbers, key=lambda x: str(x)*3, reverse=True)
    stringified_numbers = list(map(str, sorted_numbers))
    
    return '0' if stringified_numbers[0] == '0' else ''.join(stringified_numbers)